### PR TITLE
EOS-26043: Implement Expiration time filter for messages for a particular topic

### PR DESCRIPTION
**Pull request by suryakumar1024**
# Problem Statement
- A message type should have an expiration time, such that if a message was not read within the expiration time, it should get removed from message bus.
- mb_admin.set_message_type_expire('message_type', time_to_expire_in_milli_sec)

# Design
-  For Bug describe the fix here. 
-  For Feature, Post the link to the solution page on the confluence CORTX Foundation Library 

# Coding 
-  Coding conventions are followed and code is consistent [Y/N]: 
-  Confirm All CODACY errors are resolved [Y/N]: 

# Testing 
- [x] Confirm that Test Cases are added (for both the cases, fix and feature) [Y/N]: 
- [x] Confirm Test Cases cover Happy Path, Non-Happy Path and Scalability [Y/N]: 
- [x] Confirm Testing was performed with installed RPM [Y/N]:  

# Review Checklist 
  Before posting the PR please ensure
- [x] PR is self reviewed
- [ ] Is there a change in filename/package/module or signature [Y/N]: 
- [ ] If yes for above point, Is a notification sent to all other cortx components [Y/N]
- [x] Jira is updated
- [x] Check if the description is clear and explained. 
- [x] Check Acceptance Criterion is defined. 
- [ ] All the tests performed should be mentioned before Resolving a JIRA. 
- [ ] Verification needs to be done before marked as Closed/Verified 

# Documentation
- [ ] Changes done to WIKI / Confluence page

### CODE REVIEW

**Review by rahul-tripathi-git — APPROVED** [review]
Looks good to me.

File: py-utils/src/utils/message_bus/message_bus.py [hunk 1]
```diff
@@ -118,3 +118,11 @@ def receive(self, client_id: str, timeout: float = None) -> list:
     def ack(self, client_id: str):
         """ Provides acknowledgement on offset """
         self._broker.ack(client_id)
+
+    def set_message_type_expire(self, client_id: str, message_type: str, \
+        expire_time: int):
+        """
+        Set expiration time for given message type 
+        """
+        self._broker.set_message_type_expire(client_id,message_type, \
```
Comment (selvakumaar5496) [inline py-utils/src/utils/message_bus/message_bus.py:127]: provide space after comma

File: py-utils/src/utils/message_bus/message_bus_client.py [hunk 2]
```diff
@@ -121,6 +121,19 @@ def delete(self):
         client_id = self._get_conf('client_id')
         return self._message_bus.delete(client_id, message_type)
 
+    def set_message_type_expire(self, message_type, expire_time: int):
```
Comment (selvakumaar5496) [inline py-utils/src/utils/message_bus/message_bus_client.py:124]: specify type of message_type also

File: py-utils/src/utils/message_bus/message_bus_client.py [hunk 3]
```diff
@@ -121,6 +121,19 @@ def delete(self):
...
+        """
+        Set expiration time for given message type
+        """
+        message_type_list = self.list_message_types()
+        if message_type not in message_type_list:
+            raise MessageBusError(errno.ENOENT, "Unknow topic: Could not"+\
+            " found the topic %s in topic list %s", message_type, \
```
Comment (selvakumaar5496) [inline py-utils/src/utils/message_bus/message_bus_client.py:131]: indent to 4 table spaces

File: py-utils/test/message_bus/test_message_bus.py [hunk 4]
```diff
@@ -171,13 +171,24 @@ def test_013_purge_messages(self):
         message = TestMessageBus._consumer.receive()
         self.assertIsNone(message)
 
+    def test_014_set_message_type_expiry(self):
+        """Test set message type expiry"""
```
Comment (selvakumaar5496) [inline py-utils/test/message_bus/test_message_bus.py:175]: providing a period symbol will remove this codacy issue. Refer to other testcases

File: py-utils/src/utils/message_bus/message_bus_client.py [hunk 5]
```diff
@@ -121,6 +121,19 @@ def delete(self):
         client_id = self._get_conf('client_id')
         return self._message_bus.delete(client_id, message_type)
 
+    def set_message_type_expire(self, message_type, expire_time: int):
+        """
+        Set expiration time for given message type
+        """
+        message_type_list = self.list_message_types()
+        if message_type not in message_type_list:
+            raise MessageBusError(errno.ENOENT, "Unknow topic: Could not"+\
```
Comment (sachinpunadikar) [inline py-utils/src/utils/message_bus/message_bus_client.py:130]: Correction needed Unknow -> Unknown

**Review by selvakumaar5496 — CHANGES_REQUESTED** [review]
Please work on the comments.

File: py-utils/src/utils/message_bus/message_bus_client.py (hunk 3 above)
Comment (sachinpunadikar) [inline py-utils/src/utils/message_bus/message_bus_client.py:131]: found -> find

File: py-utils/src/utils/message_bus/message_bus_client.py [hunk 6]
```diff
@@ -121,6 +121,19 @@ def delete(self):
         client_id = self._get_conf('client_id')
         return self._message_bus.delete(client_id, message_type)
 
+    def set_message_type_expire(self, message_type, expire_time: int):
+        """
+        Set expiration time for given message type
+        """
+        message_type_list = self.list_message_types()
+        if message_type not in message_type_list:
```
Comment (selvakumaar5496) [inline py-utils/src/utils/message_bus/message_bus_client.py:129]: Do we need this validation here? We usually dont have any validation in client code.
We shall do the same in KafkaMessageBroker class. 
Please confirm @sachinpunadikar

File: py-utils/test/message_bus/test_message_bus.py [hunk 7]
```diff
@@ -171,13 +171,24 @@ def test_013_purge_messages(self):
...
+        TestMessageBus._admin.set_message_type_expire(\
+            TestMessageBus._message_type, 1)
+        TestMessageBus._producer.send(["A simple test message"])
+        message = TestMessageBus._consumer.receive()
```
Comment (sachinpunadikar) [inline py-utils/test/message_bus/test_message_bus.py:179]: In test code, one should wait for the expiry time to exceed & try to read the message. Should not get the message, also 1 more test case needed where one reads the message before expiry time.

File: py-utils/src/utils/message_bus/message_bus.py (hunk 1 above)
Comment (suryakumar1024) [reply py-utils/src/utils/message_bus/message_bus.py]: Updated in recent patch

File: py-utils/src/utils/message_bus/message_bus_client.py (hunk 5 above)
Comment (suryakumar1024) [reply py-utils/src/utils/message_bus/message_bus_client.py]: Updated in recent patch

File: py-utils/src/utils/message_bus/message_bus_client.py (hunk 2 above)
Comment (suryakumar1024) [reply py-utils/src/utils/message_bus/message_bus_client.py]: Updated in recent patch

File: py-utils/src/utils/message_bus/message_bus_client.py (hunk 3 above)
Comment (suryakumar1024) [reply py-utils/src/utils/message_bus/message_bus_client.py]: Updated in recent patch

File: py-utils/test/message_bus/test_message_bus.py (hunk 4 above)
Comment (suryakumar1024) [reply py-utils/test/message_bus/test_message_bus.py:175]: Updated in recent patch

File: py-utils/src/utils/message_bus/message_bus_client.py (hunk 3 above)
Comment (suryakumar1024) [reply py-utils/src/utils/message_bus/message_bus_client.py]: Updated in recent patch

File: py-utils/test/message_bus/test_message_bus.py (hunk 4 above)
Comment (rahul-tripathi-git) [reply py-utils/test/message_bus/test_message_bus.py:175]: looks like your changes are not update here, please check.

File: py-utils/src/utils/message_bus/message_bus.py [hunk 8]
```diff
@@ -118,3 +118,11 @@ def receive(self, client_id: str, timeout: float = None) -> list:
...
         """ Provides acknowledgement on offset """
         self._broker.ack(client_id)
+
+    def set_message_type_expire(self, client_id: str, message_type: str, \
+        expire_time: int):
+        """
+        Set expiration time for given message type 
+        """
+        self._broker.set_message_type_expire(client_id, message_type, \
+            expire_time)
```
Comment (selvakumaar5496) [inline py-utils/src/utils/message_bus/message_bus.py:128]: Make this as a single line comment and provide a period symbol(.)

File: py-utils/src/utils/message_bus/message_bus_client.py (hunk 6 above)
Comment (sachinpunadikar) [reply py-utils/src/utils/message_bus/message_bus_client.py:129]: I agree, lets move it to appropriate place.

File: py-utils/test/message_bus/test_message_bus.py (hunk 4 above)
Comment (suryakumar1024) [reply py-utils/test/message_bus/test_message_bus.py]: updated in recent patch

File: py-utils/test/message_bus/test_message_bus.py (hunk 7 above)
Comment (suryakumar1024) [reply py-utils/test/message_bus/test_message_bus.py:183]: added in recent patch

File: py-utils/src/utils/message_bus/message_broker_collection.py [hunk 9]
```diff
@@ -563,3 +563,44 @@ def ack(self, consumer_id: str):
...
+                    raise MessageBusError(errors.ERR_OP_FAILED, \
+                        "Updating message type expire time by "+\
+                        "alter_configs() for resource %s failed using admin" +\
+                        "%s for message type %s", topic_resource, admin,\
+                        message_type)
+                continue
+            else:
+                break
+        Log.debug("Successfully updated message type expire time.")
+        return 0
```
Comment (sachinpunadikar) [inline py-utils/src/utils/message_bus/message_broker_collection.py:606]: do we need this return statement ?

File: py-utils/src/utils/message_bus/message_bus.py (hunk 8 above)
Comment (suryakumar1024) [reply py-utils/src/utils/message_bus/message_bus.py:128]: Updated in recent patch

File: py-utils/test/message_bus/test_message_bus.py [hunk 10]
```diff
@@ -171,13 +171,54 @@ def test_013_purge_messages(self):
...
+        """
+        # Set expire time to 2 seconds
+        TestMessageBus._admin.set_message_type_expire(\
+            TestMessageBus._message_type, 2000)
+        TestMessageBus._producer.send(["A simple test message"])
+        # get before expire
+        message = TestMessageBus._consumer.receive()
+        self.assertEqual(message, b'A simple test message')
+
+    def test_015_message_type_read_before_expiry(self):
```
Comment (sachinpunadikar) [inline py-utils/test/message_bus/test_message_bus.py:186]: Update the test case name. Here you are trying to read after expiry of messages
test_015_message_type_read_before_expiry -> test_015_message_type_read_after_expiry

Comment (suryakumar1024) [reply py-utils/test/message_bus/test_message_bus.py]: Updated in recent patch

File: py-utils/src/utils/message_bus/message_bus.py [hunk 11]
```diff
@@ -118,3 +118,11 @@ def receive(self, client_id: str, timeout: float = None) -> list:
     def ack(self, client_id: str):
         """ Provides acknowledgement on offset """
         self._broker.ack(client_id)
+
+    def set_message_type_expire(self, client_id: str, message_type: str, \
+        expire_time: int):
+        """
```
Comment (sachinpunadikar) [inline py-utils/src/utils/message_bus/message_bus.py:124]: make this comment as single line

File: py-utils/src/utils/message_bus/message_bus.py [hunk 12]
```diff
@@ -118,3 +118,11 @@ def receive(self, client_id: str, timeout: float = None) -> list:
...
+        Set expiration time for given message type 
+        """
+        self._broker.set_message_type_expire(client_id, message_type, \
```
Comment (sachinpunadikar) [inline py-utils/src/utils/message_bus/message_bus.py:127]: check the return value here

File: py-utils/test/message_bus/test_message_bus.py [hunk 13]
```diff
@@ -171,13 +171,54 @@ def test_013_purge_messages(self):
         message = TestMessageBus._consumer.receive()
         self.assertIsNone(message)
 
+    def test_014_set_message_type_expiry(self):
+        """
```
[the remark this answers is not in the document]
Comment (suryakumar1024) [reply py-utils/test/message_bus/test_message_bus.py]: Updated in a recent patch.

File: py-utils/src/utils/message_bus/message_bus.py (hunk 12 above)
Comment (suryakumar1024) [reply py-utils/src/utils/message_bus/message_bus.py]: Updated in a recent patch.

File: py-utils/src/utils/message_bus/message_bus.py (hunk 11 above)
Comment (suryakumar1024) [reply py-utils/src/utils/message_bus/message_bus.py]: Updated in a recent patch.

**Review by sachinpunadikar — APPROVED** [review]
+1

**Review by selvakumaar5496 — APPROVED** [review]

**Review by rahul-tripathi-git — APPROVED** [review]
Looks good to me.

**Pull request merged**